iroha_p2p/src/peer.rs

```diff
@@ -29,422 +27,701 @@ use ursa::{
-        let dh = K::new();
-        let shared = match dh.compute_shared_secret(&self.secret_key, public_key) {
-            Ok(key) => key,
-            Err(e) => {
-                warn!(%e, "Error creating shared secret!");
-                return Err(Error::Keys);
-            }
-        };
-        debug!("Derived shared key: {:?}", &shared.0);
-        match self.new_encryptor(shared.0.as_slice()) {
-            Ok(encryptor) => {
-                self.cipher = Some(encryptor);
-                Ok(())
-            }
-            Err(e) => {
-                warn!(%e, "Unexpected error creating encryptor!");
-                Err(Error::Keys)
-            }
+    async fn into_peer(self) -> Addr<Peer<T, K, E>> {
+        let (r, w) = self.stream.into_split();
+        let (sender, receiver) = oneshot::channel();
+        let peer = Peer::new(
+            self.id,
+            w,
+            self.cipher.unwrap(),
+            self.network,
+            false,
+            sender,
+        );
+        let stream = Peer::<T, K, E>::stream(r, receiver);
+
+        let peer = peer.preinit();
+        let addr = peer.address.clone();
+
+        // spawning task
+        tokio::spawn(async move {
+            stream.for_each(|it| async { addr.do_send(it).await }).await;
+        });
+
+        peer.start().await
+    }
+}
+
+#[derive(iroha_actor::Message)]
+struct ContinueHandshake;
+
+#[async_trait::async_trait]
+impl<T, K, E> Actor for ToPeer<T, K, E>
+where
+    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
+    K: KeyExchangeScheme + Send + 'static,
+    E: Encryptor + Send + 'static,
+{
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        info!(addr = ?self.id.address, "Received connection from");
+        ctx.notify(ContinueHandshake);
+    }
+
+    async fn on_stop(self, _ctx: Context<Self>) {
+        self.into_peer().await;
+    }
+}
+
+#[async_trait::async_trait]
+impl<T, K, E> ContextHandler<ContinueHandshake> for ToPeer<T, K, E>
+where
+    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
+    K: KeyExchangeScheme + Send + 'static,
+    E: Encryptor + Send + 'static,
+{
+    type Result = ();
+    async fn handle(
+        &mut self,
+        ctx: &mut Context<Self>,
+        ContinueHandshake: ContinueHandshake,
+    ) -> Self::Result {
+        if self.state == ToState::Ready {
+            return;
         }
+
+        if let Err(err) = self.handshake().await {
+            info!(addr=?self.id.address, state=?self.state, %err, "Error connecting to peer");
+            ctx.stop_now();
+            return;
+        }
+        ctx.notify(ContinueHandshake);
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+enum FromState {
+    Connecting,
+    Connected,
+    SendKey,
+    GetKey,
+    Ready,
+}
+
+/// This is an endpoint, that joggles messages between [`crate::Network`] and another connected node.
+pub(crate) struct FromPeer<T, K, E: Encryptor> {
```

Comment:
The names of `ToPeer` and `FromPeer` are mismatched.

Reply:
I've swapped naming:
- `FromPeer` is used when we connect to someone
- `ToPeer` used when we got connected to

Reply:
We have two types of connections: incoming and outgoing. So, `ToPeer` is outgoing for me, and `FromPeer` is incoming for me.

Reply:
Yeah, naming is confusing. I guess we can name them Outgoing and Incoming

Reply:
Or `PeerIn` and `PeerOut` :)